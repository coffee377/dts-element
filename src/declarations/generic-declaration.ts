import * as ts from 'typescript';
import {IType} from '../collections';
import {ElementKind} from '../constants';
import {create_element, IElement} from '../element';
import {transform} from '../transform';

export interface IGenericDeclarationOptions {
  name: string;
  extends?: IType;
  defalut?: IType;
}

export interface IGenericDeclaration
  extends IElement<ElementKind.GenericDeclaration>, IGenericDeclarationOptions {}

export const create_generic_declaration = (options: IGenericDeclarationOptions): IGenericDeclaration => ({
  ...create_element(ElementKind.GenericDeclaration),
  ...options,
});

export const transform_generic_declaration = (element: IGenericDeclaration, path: IElement<any>[]) =>
  ts.createTypeParameterDeclaration(
    element.name,
    element.extends && (transform(element.extends, [...path, element]) as ts.TypeNode),
    element.defalut && (transform(element.defalut, [...path, element]) as ts.TypeNode),
  );
