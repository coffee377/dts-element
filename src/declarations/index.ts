export * from './class-declaration';
export * from './function-declaration';
export * from './generic-declaration';
export * from './interface-declaration';
export * from './parameter-declaration';
export * from './type-declaration';
export * from './variable-declaration';
