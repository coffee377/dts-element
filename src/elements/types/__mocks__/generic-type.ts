import {Type} from '../../type';

export class GenericType extends Type<any, any> {

  public get default_parameters(): any {
    return {};
  }

  public _emit(): string {
    return `[GenericType ${this.parameters.name}]`;
  }

}
