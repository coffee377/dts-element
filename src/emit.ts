import * as ts from 'typescript';
import {ElementKind} from './constants';
import {IElement} from './element';
import {emit_triple_slash_reference, ITripleSlashReference} from './others/triple-slash-reference';
import {transform} from './transform';

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) => {
  switch (element.kind) {
    case ElementKind.TripleSlashReference:
      return emit_triple_slash_reference(element as ITripleSlashReference);
    default:
      const ts_node = transform(element);
      const source_file = ts.createSourceFile('', '', ts.ScriptTarget.Latest, false, ts.ScriptKind.TS);
      const printer = ts.createPrinter(options);
      return printer.printNode(ts.EmitHint.Unspecified, ts_node, source_file);
  }
};
