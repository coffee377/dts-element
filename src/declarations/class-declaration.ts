import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {create_element, IElement, IElementOptions} from '../element';
import {IClassMember} from '../members';
import {IIndexSignature} from '../others/index-signature';
import {transform} from '../transform';
import {IGeneralType} from '../types/general-type';
import {add_declare_modifier_if_need, create_type_nodes, create_type_parameters} from '../utils';
import {IGenericDeclaration} from './generic-declaration';

export interface IClassDeclarationOptions extends IElementOptions {
  name: string;
  export?: boolean;
  generics?: IGenericDeclaration[];
  members?: (IClassMember | IIndexSignature)[];
  abstract?: boolean;
  extends?: IGeneralType;
}

export interface IClassDeclaration
  extends IElement<ElementKind.ClassDeclaration>, IClassDeclarationOptions {}

export const create_class_declaration = (options: IClassDeclarationOptions): IClassDeclaration => ({
  ...create_element(ElementKind.ClassDeclaration),
  ...options,
});

// tslint:disable:ter-indent max-line-length

export const transform_class_declaration = (element: IClassDeclaration, path: IElement<any>[]) =>
  ts.createClassDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
                            [
                              ...(element.export === true)
                                ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
                                : [],
                              ...(element.abstract === true)
                                ? [ts.createToken(ts.SyntaxKind.AbstractKeyword)]
                                : [],
                            ],
                            path,
                          ),
    /* name            */ element.name,
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* heritageClauses */ (element.extends === undefined)
                            ? []
                            : [ts.createHeritageClause(
                            /* token */ ts.SyntaxKind.ExtendsKeyword,
                            /* types */ [ts.createExpressionWithTypeArguments(
                                          /* typeArguments */ create_type_nodes(element.extends.generics, path),
                                          /* expression    */ ts.createIdentifier(element.extends.name),
                                        )],
                            )],
    /* members         */ (element.members || []).map(member => transform(member, path) as ts.ClassElement),
  );
