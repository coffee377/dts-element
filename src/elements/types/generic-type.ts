import {emit_equal} from '../../helpers/emit-equal';
import {emit_extends} from '../../helpers/emit-extends';
import {Stack} from '../../stack';
import {InterfaceDeclaration} from '../declarations/interface-declaration';
import {Type} from '../type';

export interface IGenericTypeRequiredParameters {
  name: string;
}

export interface IGenericTypeOptionalParameters {
  extends: Type | null;
  default: Type | null;
}

/**
 * ```ts
 * type X<T extends string> = T;
 * //     ^^^^^^^^^^^^^^^^    ^
 * ```
 */
export class GenericType extends Type<IGenericTypeRequiredParameters, IGenericTypeOptionalParameters> {

  private _instance_of_generic_type: true;

  public get default_parameters(): IGenericTypeOptionalParameters {
    return {
      extends: null,
      default: null,
    };
  }

  public _emit(stack: Stack): string {
    const {name} = this.parameters;
    const generic_extends = emit_extends(this.parameters.extends, stack);
    const generic_default = emit_equal(this.parameters.default, stack);
    return `${name}${generic_extends}${generic_default}`;
  }

}
