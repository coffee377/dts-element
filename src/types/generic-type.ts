import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {IGenericDeclaration} from '../declarations/generic-declaration';
import {create_element, IElement} from '../element';
import {transform_general_type} from './general-type';

export interface IGenericTypeOptions {
  name: string | IGenericDeclaration;
}

export interface IGenericType
  extends IElement<ElementKind.GenericType>, IGenericTypeOptions {}

export const create_generic_type = (options: IGenericTypeOptions): IGenericType => ({
  ...create_element(ElementKind.GenericType),
  ...options,
});

export const transform_generic_type: (element: IGenericType, path: IElement<any>[]) => ts.TypeReferenceNode
  = transform_general_type as any;
