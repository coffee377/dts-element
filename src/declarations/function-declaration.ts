import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {create_element, IElement} from '../element';
import {transform} from '../transform';
import {create_function_type, IFunctionType} from '../types/function-type';

export interface IFunctionDeclarationOptions {
  name: string | undefined;
  type?: IFunctionType;
  export?: boolean;
}

export interface IFunctionDeclaration
  extends IElement<ElementKind.FunctionDeclaration>, IFunctionDeclarationOptions {}

export const create_function_declaration = (options: IFunctionDeclarationOptions): IFunctionDeclaration => ({
  ...create_element(ElementKind.FunctionDeclaration),
  ...options,
});

export const transform_function_declaration = (element: IFunctionDeclaration, path: IElement<any>[]) => {
  const function_type = transform(element.type || create_function_type(), path) as ts.FunctionDeclaration;
  return ts.createFunctionDeclaration(
    /* decorators     */ undefined,
    /* modifiers      */ (element.export === true)
                           ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
                           : undefined,
    /* asteriskToken  */ undefined,
    /* name           */ element.name,
    /* typeParameters */ function_type.typeParameters,
    /* parameters     */ function_type.parameters,
    /* type           */ function_type.type,
    /* body           */ undefined,
  );
};
