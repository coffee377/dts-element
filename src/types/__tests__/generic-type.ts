import {create_generic_declaration} from '../../declarations/generic-declaration';
import {emit} from '../../emit';
import {create_generic_type} from '../generic-type';

it('should return correctly with name', () => {
  expect(emit(
    create_generic_type({
      name: 'T',
    }),
  )).toMatchSnapshot();
});
