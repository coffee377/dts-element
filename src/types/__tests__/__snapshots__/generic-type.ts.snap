// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly with name 1`] = `"T"`;
